[shopsys] Unified using ENVs (#2040)
shopsys/shopsys: docs/introduction/setting-environment-variables.md

# Setting environment variables

In this article describes how to set environments variable in different environment.

The most common uses are: Kubernetes, Docker or native installation.

!!! note
    

>     Bellow is descibed the easiest way. For better understanding we recomend to find out the documentation of the specific environment.

Reviewer: I'm afraid this could be mixed up with the application environment (prod/test/dev).

I would refer Docker/Native/K8S as a `platform` or similar word

!!! tip

>     You may want to tune your configuration via [`.env` file](https://symfony.com/doc/4.4/configuration.html#configuring-environment-variables-in-env-files).

Reviewer: ```suggestion
    You may want to set your configuration via [`.env` file](https://symfony.com/doc/4.4/configuration.html#configuring-environment-variables-in-env-files).
```

Author: this whole tip was removed

>     You may want to tune your configuration via [`.env` file](https://symfony.com/doc/4.4/configuration.html#configuring-environment-variables-in-env-files).

Reviewer: I suggest mentioning `.env.local` file as it's not committed and may come in handy on the local machine.

Author: mentioned

+           kind: Deployment
+           name: webserver-php-fpm
+       path: ./webserver-php-fpm-patch.yaml
configMapGenerator:
    -   name: nginx-configuration
```

- create `kubernetes/kustomize/overlays/ci/webserver-php-fpm-patch.yaml` containing

```diff
+-   op: add
+    path: /spec/template/spec/containers/0/env/-
+    value:
+        name: REDIS_PREFIX
+        value: 'my_awesome_app'
```

## Docker

When using docker containers without kubernetes add the environment variable to the `docker-compose.yml` file to `php-fpm` definition like in example below

```diff
    php-fpm:
        build:
            context: .
            dockerfile: docker/php-fpm/Dockerfile
            target: development
            args:
                www_data_uid: 1000
                www_data_gid: 1000
        container_name: shopsys-framework-php-fpm
        volumes:
            - shopsys-framework-sync:/var/www/html
            - shopsys-framework-vendor-sync:/var/www/html/vendor
            - shopsys-framework-web-sync:/var/www/html/web
        ports:
            - "35729:35729"
+       environment:
+           - REDIS_PREFIX='my_awesome_app'
```

## Native instalation

> Without containers you must set environment variable on the host machine, typically in unix like OS by executing

Reviewer: I would add info about the possibility to set the environment variable with the `.env` and `.env.local` file.
It's the easiest way when installed natively.

Author: This is really good point. I updated the whole article as the .env files are commonly used directly by symfony.

```

> export REDIS_PREFIX='my_awesome_app'

Reviewer: It would be nice to explain the priority of environment variables. 

Priority is 
System env > .env.local file > .env file > parameters.yaml value

Author: Sure. I have not described by myslef but linked to Symfony doc.

# Setting environment variables

This article describes how to set configure environment variables in different environment and even on some common used webservers.

## Global configuration of environment variables

Shopsys Framework as Symfony based application is using the same environment configuration as Symfony.

> We recommend using [`.env` files](https://symfony.com/doc/4.4/configuration.html#configuring-environment-variables-in-env-files) for configuring environment variables as it can be easily maintained. There is also advantage of different configuration through different application environments (suh as dev, test and prod).

Reviewer: ```suggestion
We recommend using [`.env` files](https://symfony.com/doc/4.4/configuration.html#configuring-environment-variables-in-env-files) for configuring environment variables as it can be easily maintained. There is also advantage of different configuration through different application environments (such as dev, test and prod).
```

!!! note

    You may also use [`.env.local` file](https://symfony.com/doc/4.4/configuration.html#overriding-environment-values-via-env-local) for your personal configuration. The `*.local` files should not be commited so they are ingored with `.gitignore` by default.

For better understanding [visit the whole documentation here](https://symfony.com/doc/4.4/configuration.html#configuration-environments) where you can find more info about priority, [variables encrypting](https://symfony.com/doc/4.4/configuration.html#encrypting-environment-variables-secrets), [syntax](https://symfony.com/doc/4.4/configuration.html#env-file-syntax) and much more.

## Overriding by real environment variable of webserver

> When `.env` files are not enough for your purpose you may override environment variables directly by webserver setting. These settings ignore any variables configured in `.env` files.

Reviewer: ```suggestion
When `.env` files are not enough for your purpose you may override environment variables directly by webserver setting. These settings override any variables configured in `.env` files.
```

The most common uses are: Kubernetes, Docker or native installation.

!!! note
    

>     Bellow is descibed the easiest way. For better understanding we recomend to find out the documentation of the specific platform.

Reviewer: ```suggestion
    Bellow is described the easiest way. For better understanding, we recommend finding out the documentation of the specific platform.
```

### Kubernetes

When you are using kubernetes on CI server change your configuration of:

- `kubernetes/kustomize/overlays/ci/kustomization.yaml`

```diff
        path: ./ingress-patch.yaml
+   -   target:
+           group: apps
+           version: v1
+           kind: Deployment
+           name: webserver-php-fpm
+       path: ./webserver-php-fpm-patch.yaml
configMapGenerator:
    -   name: nginx-configuration
```

- create `kubernetes/kustomize/overlays/ci/webserver-php-fpm-patch.yaml` containing

```diff
+-   op: add
+    path: /spec/template/spec/containers/0/env/-
+    value:
+        name: REDIS_PREFIX
+        value: 'my_awesome_app'
```

### Docker

> When using docker containers without kubernetes add the environment variable to the `docker-compose.yml` file to `php-fpm` definition like in example below

Reviewer: ```suggestion
When using docker containers without kubernetes, add the environment variable to the `docker-compose.yml` file to `php-fpm` definition like in the example below
```
